ksqldb-engine/src/main/java/io/confluent/ksql/engine/KsqlEngine.java

```diff
@@ -293,6 +319,195 @@ public TransientQueryMetadata executeTransientQuery(
     }
   }
 
+  /**
+   * Unlike the other queries, stream pull queries are split into create and wait because the three
+   * API endpoints all need to do different stuff before, in the middle of,
+   * and after these two phases. One of them actually needs to wait on the pull
+   * query in a callback after starting the query, so splitting it into two method calls was
+   * the most practical choice.
+   */
+  public TransientQueryMetadata createStreamPullQuery(
+      final ServiceContext serviceContext,
+      final ConfiguredStatement<Query> statementOrig,
+      final boolean excludeTombstones
+  ) {
+
+    if (!ksqlConfig.getBoolean(KsqlConfig.KSQL_QUERY_STREAM_PULL_QUERY_ENABLED)) {
+      throw new KsqlStatementException(
+          "Pull queries on streams are disabled. To create a push query on the stream,"
+              + " add EMIT CHANGES to the end. To enable pull queries on streams, set"
+              + " the " + KsqlConfig.KSQL_QUERY_STREAM_PULL_QUERY_ENABLED + " config to 'true'.",
+          statementOrig.getStatementText()
+      );
+    }
+
+    // Stream pull query overrides: start from earliest, use one  thread, and use a tight commit
+    // interval for responsiveness.
+    // Starting from earliest is semantically necessary.
+    // Using a single thread keeps these queries as lightweight as possible, since we are
+    // not counting them against the transient query limit.
+    // Setting the commit interval to 100ms to make the query results snappier.
+    // Since this is a "sit and wait", not a "fire and forget", query,
+    // we want to make sure people are going to see their results come back asap.
+    final ConfiguredStatement<Query> statement = statementOrig.withConfigOverrides(
+        ImmutableMap.<String, Object>builder()
+            .putAll(statementOrig.getSessionConfig().getOverrides())
+            .put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+            .put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 1)
+            .put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100)
+            .build()
+    );
+    try {
+
+      final TransientQueryMetadata transientQueryMetadata = EngineExecutor
+          .create(primaryContext, serviceContext, statement.getSessionConfig())
+          .executeTransientQuery(statement, excludeTombstones);
+
+      QueryLogger.info(
+          "Streaming stream pull query results '{}'",
+          statement.getStatementText()
+      );
+
+      return transientQueryMetadata;
+
+    } catch (final KsqlStatementException e) {
```

Comment:
For what it's worth, I think KsqlStatementException is taken as a 4XX for at least `/query-stream`.  Making things consistent is probably part of a larger fix, but just wanted to point that out here since all exceptions are wrapped in it.

Reply:
I see. In retrospect, it probably makes more sense just to rethrow KsqlExceptions up to the API layer so it can return either a 4xx or a 5xx.